features/storage/kvstore/KVStore.h

```diff
@@ -0,0 +1,199 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_KVSTORE_H
+#define MBED_KVSTORE_H
+
+#include <stdint.h>
+#include <stdio.h>
+
+typedef enum {
+    KVSTORE_SUCCESS                =  0,
+    KVSTORE_READ_ERROR             = -1,
+    KVSTORE_WRITE_ERROR            = -2,
+    KVSTORE_NOT_FOUND              = -3,
+    KVSTORE_DATA_CORRUPT           = -4,
+    KVSTORE_BAD_VALUE              = -5,
+    KVSTORE_NO_SPACE_ON_DEVICE     = -6,
+    KVSTORE_OS_ERROR               = -7,
+    KVSTORE_WRITE_ONCE_ERROR       = -8,
+    KVSTORE_AUTH_ERROR             = -9,
+    KVSTORE_RBP_AUTH_ERROR         = -10,
+    KVSTORE_MAX_KEYS_REACHED       = -11,
+    KVSTORE_UNINITIALIZED          = -12,
+} kvstore_status_e;
+
+
+
+/** KVStore class
+ *
+ *  Interface class for Key Value Storage
+ */
+
+class KVStore {
+public:
+    enum create_flags {
+        WRITE_ONCE_FLAG       = (1 << 0),
+        ENCRYPT_FLAG          = (1 << 1),
+        AUTHENTICATE_FLAG     = (1 << 2),
+        ROLLBACK_PROTECT_FLAG = (1 << 3),
+    };
+
+    static const uint32_t MAX_KEY_SIZE = 128;
+
+    typedef struct _opaque_set_handle *set_handle_t;
+
+    typedef struct _opaque_key_iterator *iterator_t;
+
+    typedef struct info {
+        size_t size;
+        uint32_t flags;
+    } info_t;
+
+    virtual ~KVStore() {};
+
+    /**
+     * @brief Initialize KVStore
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int init() = 0;
+
+    /**
+     * @brief Deinitialize KVStore
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int deinit() = 0;
+
+
+    /**
+     * @brief Reset KVStore contents (clear all keys)
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int reset() = 0;
+
+    /**
+     * @brief Set one KVStore item, given key and value.
+     *
+     * @param[in]  key                  Key.
+     * @param[in]  buffer               Value data buffer.
+     * @param[in]  size                 Value data size.
+     * @param[in]  create_flags         Flag mask.
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int set(const char *key, const void *buffer, size_t size, uint32_t create_flags);
+
+    /**
+     * @brief Get one KVStore item, given key.
+     *
+     * @param[in]  key                  Key.
+     * @param[in]  buffer               Value data buffer.
+     * @param[in]  buffer_size          Value data buffer size.
+     * @param[out] actual_size          Actual read size.
+     * @param[in]  offset               Offset to read from in data.
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int get(const char *key, void *buffer, size_t buffer_size, size_t *actual_size = NULL, size_t offset = 0);
```

Comment:
Here's an interesting thing I just found out.

Default arguments are not easily optimized out by compilers. The way they work is unfortunately hard for compilers to work around : (

Consider the case where a get with an offset is expensive, but a get with default offset is cheap (in terms of code size). Default arguments aren't different functions, so can't be gced by the compiler. The other trick a compiler can do, constant propagation, can't kick in unless LTO is on because this is a non-static function. So even if a user never uses a non-zero offset, they will still have to pay for all the offset logic.

I don't know if offset is expensive or not, but it's something to keep in mind. If a default is expensive it's probably better as an additional overloaded function.

And unfortunately for interface classes, the interface has to make the decision between overloads vs defaults, the implementation class can't add virtual functions easily.